# Exploring Ebay Car Sales Data

This is a guided project on the Data Analyst's Path from Dataquest. The main focus is to practice Data Cleaning using Pandas in Jupyter Notebook.

The dataset to be cleaned and analyzed has Ebay Car Sales Data from Ebay Germany. The dataset was edited by Dataquest's team to match more realistically a web scraped dataset. The columns, when strings, are written in german.

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date;
- name - Name of the car;
- seller - Whether the seller is private or a dealer;
- offerType - The type of listing;
- price - The price on the ad to sell the car;
- abtest - Whether the listing is included in an A/B test;
- vehicleType - The vehicle Type;
- yearOfRegistration - The year in which the car was first registered;
- gearbox - The transmission type;
- powerPS - The power of the car in PS;
- model - The car model name;
- kilometer - How many kilometers the car has driven;
- monthOfRegistration - The month in which the car was first registered;
- fuelType - What type of fuel the car uses;
- brand - The brand of the car;
- notRepairedDamage - If the car has a damage which is not yet repaired;
- dateCreated - The date on which the eBay listing was created;
- nrOfPictures - The number of pictures in the ad;
- postalCode - The postal code for the location of the vehicle;
- lastSeenOnline - When the crawler saw this ad last online.

#### Importing libraries

In [1]:
import numpy as np
import pandas as pd
import datetime as dt

#### Initial exploration of the csv file

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

#### First observations

The column names use *camelCase* instead of Python's preferred *snake_case*, which means we can't just replace spaces with underscores, so change is needed in this matter. The columns `price` and `odometer` need to read `int` values instead of the usual DataFrame object. To do this, we need to remove extra characters and then change the column type. The column `nrOfPictures` only has the value of 0, so it's wise to delete it, since it won't help our analysis. The same observation can be made to the columns `seller` and `offerType`, since they only have the individual values of *privat* and *Angebot*, respectively. 

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

#### Converting columns to *snake_case*

In [6]:
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.columns = new_columns

autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

#### Exploring the `price` and `odometer` columns

In [7]:
autos['price'] = (autos['price']
                         .str.replace("$","")
                         .str.replace(",","")
                         .astype(int)
                 )

autos['odometer'] = (autos['odometer']
                         .str.replace("km","")
                         .str.replace(",","")
                         .astype(int)
                 )

autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

#### Deleting the `nrOfPictures`, `seller` and `offerType` columns

In [374]:
autos = autos.drop(["nr_of_pictures", "seller", "offer_type"], 1)

autos.head(10)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   price   abtest vehicle_type  registration_year    gearbox  power_PS  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   
5   7900     test          bus               2006  automatik       150   
6    300     test    limousine               1995    manuell        90   
7   1990  control    limousine               1998    manuell        90   
8    250     test          NaN               2000    manuell         0   
9    590  control          bus               1997    manuell        90   

     model  odometer_km  registration_month fuel_type       brand  \
0   andere       150000                   3       lpg     peugeot   
1      7er       150000                   6    benzin         bmw   
2     golf        70000                   7    benzin  volkswagen   
3   fortwo        70000                   6    benzin       smart   
4    focus       150000                   7    benzin        ford   
5  voyager       150000                   4    diesel    chrysler   
6     golf       150000                   8    benzin  volkswagen   
7     golf       150000                  12    diesel  volkswagen   
8    arosa       150000                  10       NaN        seat   
9   megane       150000                   7    benzin     renault   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
5               NaN  2016-03-21 00:00:00        22962  2016-04-06 09:45:21  
6               NaN  2016-03-20 00:00:00        31535  2016-03-23 02:48:59  
7              nein  2016-03-16 00:00:00        53474  2016-04-07 03:17:32  
8              nein  2016-03-22 00:00:00         7426  2016-03-26 18:18:10  
9              nein  2016-03-16 00:00:00        15749  2016-04-06 10:46:35

#### Analyzing the values of the `price` column

We can see that the values of the `price` column goes from 0 euros to almost 100 million euros, which is odd, at least. This motivates a more detailed study on what's happening in this column.

In [376]:
autos['price'].value_counts().sort_index(ascending=True)

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

In [378]:
autos.sort_values(by="price", ascending=False).head(20)

date_crawled                                               name  \
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
36818  2016-03-27 18:37:37                                        Porsche_991   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   

          price   abtest vehicle_type  registration_year    gearbox  power_PS  \
39705  99999999  control    limousine               1999  automatik       224   
42221  27322222  control    limousine               2014    manuell       163   
27371  12345678  control          NaN               2017        NaN        95   
39377  12345678  control          NaN               2018    manuell        95   
47598  12345678  control    limousine               2001    manuell       101   
2897   11111111     test    limousine               1973    manuell        48   
24384  11111111     test          NaN               1995        NaN         0   
11137  10000000  control        coupe               1960    manuell       368   
47634   3890000     test        coupe               2006        NaN       799   
7814    1300000  control        coupe               1992        NaN         0   
22947   1234566  control        kombi               1999  automatik       190   
43049    999999     test          bus               1981    manuell        70   
514      999999     test        kombi               2009    manuell       101   
37585    999990     test    limousine               1985    manuell       111   
36818    350000  control        coupe               2016    manuell       500   
14715    345000  control       cabrio               2012  automatik       460   
34723    299000     test        coupe               1977    manuell       260   
35923    295000     test       cabrio               2015  automatik       400   
12682    265000  control        coupe               2016  automatik       500   
47337    259000     test       cabrio               2001    manuell       400   

             model  odometer_km  registration_month fuel_type           brand  \
39705     s_klasse       150000                   9    benzin   mercedes_benz   
42221           c4        40000                   2    diesel         citroen   
27371        punto       150000                   0       NaN            fiat   
39377          v40       150000                   6       NaN           volvo   
47598       vectra       150000                   3    benzin            opel   
2897        escort        50000                   3    benzin            ford   
24384          NaN       150000            

Since eBay also works with auctionlike sales, we will maintain prices from 1 euro up to 350.000 euros, because above this value, unrealistic prices are listed.

In [8]:
autos = autos[autos["price"].between(1,350000)]

#### Analyzing the `odometer_km` column

In [380]:
autos["odometer_km"].value_counts()

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

The `odometer_km` column only has 13 unique values, which probably represents the aproximated mileage of the vehicle chosen by multiple choice. No row has impossible values, so no further changes are needed in this column.

#### Exploring columns that contain dates

There are 5 columns on the dataset that contain date: `date_crawled`, `registration_year`, `registration_month`, `ad_created` and `last_seen`. The `registration_year` and `registration_month` are `int64` columns, while the others are `DataFrame object` columns.

#### Exploring `DataFrame object` date columns

In [381]:
autos[['date_crawled', 'ad_created', 'last_seen']].head(5)

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The dates at these columns are formatted as: **Year/Month/Day Hour:Minute:Second**. Besides, the `ad_created` column has all its hour, minute and second data with the 0 value on the first 5 rows of data.

In [382]:
autos['ad_created'].str[-8:].value_counts()


00:00:00    48565
Name: ad_created, dtype: int64

This behavior is repeated a lot throughout the dataset, so it's better to delete this, as it doesn't seem accurate. 

In [383]:
autos['ad_created'] = autos['ad_created'].str[:-9]

In [384]:
autos[['date_crawled', 'ad_created', 'last_seen']].head(5)

date_crawled  ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01  2016-04-01 14:38:50

In [385]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

We can see that the acquired data goes from March 5th of 2016 and April 7th of 2016, being April 3rd the day with the most sales data.

In [386]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

Along the `ad_created` column, we can see that the most recent ad created is from April 7th of 2016, but the oldest one is way older than the period of data collection. We can see that most ads were created at the period of data collection, though.

In [387]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

This column informs us the date that each ad from data has been last seen. We can see a near uniform distribution until April 4th. On the last 3 days, though, we can observe a significant increase of these values, becoming from 6 to 10 times bigger than the average of before.

It is likely that one add is last seen just before its closure, due to a concluded sale. This explains the inflated results of the last 3 days: as we only gathered data until April 7th, those last 3 days contain not only ads last seen due to its sale, but also views from the sale process. That means that a big chunk of these ads are yet to be seen in a near future not included in the analysis.

#### Exploring `int64` columns

The `int64` columns are `registration_year` e `registration_month`.

In [388]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Starting with the `registration_year` column, we can see some strange values, like a year 1000 of car registration and a year 9999. We can see all values of this column below.

In [389]:
autos['registration_year'].value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

In [390]:
(~autos['registration_year'].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

We removed ads that contained car's registration years outside the period from 1900 and 2016, that represented 4% of total data obtained.

In [391]:
autos = autos[autos['registration_year'].between(1900,2016)]

Now, exploring the `registration_month` column.

In [392]:
autos['registration_month'].describe()

count    46681.000000
mean         5.827125
std          3.670300
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [393]:
autos['registration_month'].value_counts(normalize=True).sort_index()

0     0.086245
1     0.066515
2     0.061053
3     0.103640
4     0.083417
5     0.083053
6     0.088237
7     0.079733
8     0.064716
9     0.069472
10    0.074870
11    0.069172
12    0.069879
Name: registration_month, dtype: float64

Na coluna `registration_month`, we can see the presence of the unexistent month 0. This can mean the announcer doesn't remember the actual month, or maybe doesn't know it. Since it represents 9% of total data, being the second most common value, it's not wise to delete these rows, so we'll keep them as they are.

#### Exploring information by brands

We want to get different informations by brand of car. Which brand is the most expensive one? Are there differences between the average registration years from each different brand? These are the two questions we are going to answer.

In [394]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

We will analyze just brands that correspond to at least 5% of the whole data collected.

In [395]:
brand_counts = autos['brand'].value_counts(normalize=True)
brands = brand_counts[brand_counts > 0.05].index
brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [396]:
brand_avg_price = {}
brand_avg_year = {}

for brand in brands:
    brandnow = autos[autos['brand'] == brand]
    avg_price = brandnow['price'].sum()/brandnow.shape[0]
    avg_year = brandnow['registration_year'].sum()/brandnow.shape[0]
    brand_avg_price[brand] = avg_price
    brand_avg_year[brand] = avg_year
    
    
print(brand_avg_price)
print(brand_avg_year)


{'volkswagen': 5402.410261610221, 'bmw': 8332.820517811953, 'opel': 2975.2419354838707, 'mercedes_benz': 8628.450366422385, 'audi': 9336.687453600594, 'ford': 3749.4695065890287}
{'volkswagen': 2002.4672480227134, 'bmw': 2003.0352345727079, 'opel': 2002.2232178414974, 'mercedes_benz': 2002.031312458361, 'audi': 2004.1559020044544, 'ford': 2002.7698437021147}


Regardless of the brand, cars announced on eBay at the period of analysis are, on average, from the years 2002 to 2004. It is noticeable that cars from BMW, Mercedez-Benz and Audi has higher average prices in comparison to the other three brands, with approximately 8.000 euros, which is natural, due to the fact that those are luxury brands. Volkswagen comes in 4th in regard to average price, with a little less than 6.000 euros. Ford and Opel are the cheapest brands analyzed, with average prices below 4.000 and 3.000, respectively.

#### Exploring mileage

We will analyse the 6 brands above to check if there is any relation between their average mileage to their average price.

In [397]:
avg_km = {}

for brand in brands:
    brandnow = autos[autos['brand'] == brand]
    avg_km_now = brandnow['odometer_km'].sum()/brandnow.shape[0]
    avg_km[brand] = avg_km_now
    
mean_price = pd.Series(brand_avg_price)    
mean_km = pd.Series(avg_km)

brand_info = pd.DataFrame(mean_price, columns=['mean_price'])

brand_info['mean_km'] = mean_km

brand_info.sort_values('mean_km', ascending=False)

mean_price        mean_km
bmw            8332.820518  132572.513140
mercedes_benz  8628.450366  130788.363313
opel           2975.241935  129310.035842
audi           9336.687454  129157.386785
volkswagen     5402.410262  128707.158791
ford           3749.469507  124266.012872

The average mileages from each brand are very similar to each other, meaning there is no reason to believe there is a direct relation between this value to their average prices.